File: app/sources.py
from dataclasses import dataclass
import json
import re
import httpx


@dataclass(frozen=True)
class Source:
    name: str
    url: str
    protocol: str


SOURCES = [
    Source("proxyscrape-socks5-fast", "https://api.proxyscrape.com/v4/free-proxy-list/get?request=displayproxies&protocol=socks5&timeout=1000&country=all&ssl=all&anonymity=all", "socks5"),
    Source("proxyscrape-http-fast", "https://api.proxyscrape.com/v4/free-proxy-list/get?request=displayproxies&protocol=http&timeout=1000&country=all&ssl=all&anonymity=all", "http"),
    Source("proxy-list-download-socks5", "https://www.proxy-list.download/api/v1/get?type=socks5", "socks5"),
    Source("proxy-list-download-http", "https://www.proxy-list.download/api/v1/get?type=http", "http"),
    Source("geonode-socks5", "https://proxylist.geonode.com/api/proxy-list?limit=500&page=1&sort_by=lastChecked&sort_type=desc&protocols=socks5", "socks5"),
    # Independently maintained, frequently refreshed public snapshots.
    Source("proxifly-http", "https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/protocols/http/data.txt", "http"),
    Source("proxifly-https", "https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/protocols/https/data.txt", "https"),
    Source("proxifly-socks4", "https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/protocols/socks4/data.txt", "socks4"),
    Source("proxifly-socks5", "https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/protocols/socks5/data.txt", "socks5"),
    Source("iplocate-http", "https://raw.githubusercontent.com/iplocate/free-proxy-list/main/protocols/http.txt", "http"),
    Source("iplocate-https", "https://raw.githubusercontent.com/iplocate/free-proxy-list/main/protocols/https.txt", "https"),
    Source("iplocate-socks4", "https://raw.githubusercontent.com/iplocate/free-proxy-list/main/protocols/socks4.txt", "socks4"),
    Source("iplocate-socks5", "https://raw.githubusercontent.com/iplocate/free-proxy-list/main/protocols/socks5.txt", "socks5"),
    Source("jetkai-all", "https://raw.githubusercontent.com/jetkai/proxy-list/main/online-proxies/txt/proxies.txt", "http"),
]


async def fetch_source(source: Source) -> set[str]:
    async with httpx.AsyncClient(timeout=15, follow_redirects=True) as client:
        response = await client.get(source.url)
        response.raise_for_status()
        text = response.text
    found = set()
    # GeoNode and similar APIs return JSON; plain-list sources return one host:port per line.
    try:
        payload = json.loads(text)
        rows = payload.get("data", payload) if isinstance(payload, dict) else payload
        for row in rows if isinstance(rows, list) else []:
            if isinstance(row, dict) and row.get("ip") and row.get("port"):
                found.add(f"{source.protocol}://{row['ip']}:{row['port']}")
        if found:
            return found
    except (ValueError, TypeError):
        pass
    for line in text.splitlines():
        value = line.strip()
        if value and re.fullmatch(r"[^:\s]+:\d{1,5}", value):
            host, port = value.rsplit(":", 1)
            if host and port.isdigit() and 1 <= int(port) <= 65535:
                found.add(f"{source.protocol}://{host}:{port}")
    return found

File: src/sources.js
import axios from 'axios';
import { setTimeout as delay } from 'node:timers/promises';
import { settings } from './config.js';
import { errorCode, isResourceError, networkGuard } from './network.js';

export const SOURCES = [
  ['proxyscrape-socks5-fast', 'https://api.proxyscrape.com/v4/free-proxy-list/get?request=displayproxies&protocol=socks5&timeout=1000&country=all&ssl=all&anonymity=all', 'socks5'],
  ['proxyscrape-http-fast', 'https://api.proxyscrape.com/v4/free-proxy-list/get?request=displayproxies&protocol=http&timeout=1000&country=all&ssl=all&anonymity=all', 'http'],
  ['geonode-socks5', 'https://proxylist.geonode.com/api/proxy-list?limit=500&page=1&sort_by=lastChecked&sort_type=desc&protocols=socks5', 'socks5'],
  ['proxifly-http', 'https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/protocols/http/data.txt', 'http'],
  ['proxifly-https', 'https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/protocols/https/data.txt', 'https'],
  ['proxifly-socks4', 'https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/protocols/socks4/data.txt', 'socks4'],
  ['proxifly-socks5', 'https://cdn.jsdelivr.net/gh/proxifly/free-proxy-list@main/proxies/protocols/socks5/data.txt', 'socks5'],
  ['iplocate-http', 'https://raw.githubusercontent.com/iplocate/free-proxy-list/main/protocols/http.txt', 'http'],
  ['iplocate-https', 'https://raw.githubusercontent.com/iplocate/free-proxy-list/main/protocols/https.txt', 'https'],
  ['iplocate-socks4', 'https://raw.githubusercontent.com/iplocate/free-proxy-list/main/protocols/socks4.txt', 'socks4'],
  ['iplocate-socks5', 'https://raw.githubusercontent.com/iplocate/free-proxy-list/main/protocols/socks5.txt', 'socks5'],
  ['thespeedx-http', 'https://raw.githubusercontent.com/TheSpeedX/SOCKS-List/master/http.txt', 'http'],
  ['thespeedx-socks4', 'https://raw.githubusercontent.com/TheSpeedX/SOCKS-List/master/socks4.txt', 'socks4'],
  ['thespeedx-socks5', 'https://raw.githubusercontent.com/TheSpeedX/SOCKS-List/master/socks5.txt', 'socks5'],
  ['monosans-http', 'https://raw.githubusercontent.com/monosans/proxy-list/main/proxies/http.txt', 'http'],
  ['monosans-socks4', 'https://raw.githubusercontent.com/monosans/proxy-list/main/proxies/socks4.txt', 'socks4'],
  ['monosans-socks5', 'https://raw.githubusercontent.com/monosans/proxy-list/main/proxies/socks5.txt', 'socks5'],
];

export function parseSourceText(source, text) {
  const found = new Set();
  try {
    const payload = JSON.parse(text);
    const rows = Array.isArray(payload) ? payload : payload?.data;
    if (Array.isArray(rows)) {
      for (const row of rows) {
        if (row?.ip && row?.port) found.add(`${source[2]}://${row.ip}:${row.port}`);
      }
      if (found.size) return found;
    }
  } catch {}

  for (const line of text.split(/\r?\n/)) {
    let value = line.trim();
    if (!value) continue;
    if (value.includes('://')) value = value.split('://', 2)[1];
    const match = value.match(/^([^:\s]+):(\d{1,5})$/);
    if (match && Number(match[2]) >= 1 && Number(match[2]) <= 65535) found.add(`${source[2]}://${match[1]}:${match[2]}`);
  }
  return found;
}

export async function fetchSource(source, client = axios, { signal } = {}) {
  let lastError;
  for (let attempt = 0; attempt < 2; attempt += 1) {
    signal?.throwIfAborted();
    if (networkGuard.paused()) throw Object.assign(new Error('Local network cooldown'), { code: 'NETWORK_COOLDOWN' });
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 15000);
    try {
      const response = await client.get(source[1], { signal: signal ? AbortSignal.any([signal, controller.signal]) : controller.signal, timeout: 15000, maxContentLength: 10 * 1024 * 1024, responseType: 'text' });
      return parseSourceText(source, response.data);
    } catch (error) {
      lastError = error;
      signal?.throwIfAborted();
      const code = errorCode(error);
      networkGuard.fail(code, settings.checkConcurrency);
      if (isResourceError(code)) throw error;
      if (attempt === 0) await delay(500, undefined, { signal });
    } finally {
      clearTimeout(timer);
      controller.abort();
    }
  }
  throw lastError;
}

export async function fetchAllSources({ signal } = {}) {
  const results = new Array(SOURCES.length);
  let next = 0;
  const worker = async () => {
    while (true) {
      const index = next++;
      if (index >= SOURCES.length) return;
      try {
        results[index] = { status: 'fulfilled', value: await fetchSource(SOURCES[index], axios, { signal }) };
      } catch (reason) {
        results[index] = { status: 'rejected', reason };
      }
    }
  };
  await Promise.all(Array.from({ length: Math.min(settings.sourceFetchConcurrency, SOURCES.length) }, worker));
  return results.map((result, index) => ({ source: SOURCES[index], result }));
}
